course3/homework-list/linked-list.h

```diff
@@ -0,0 +1,180 @@
+
+#include <cstddef>
+
+template <typename TNode>
+class LinkedList
+{
+    struct Node 
+    {
+        Node(const TNode& data, Node* next) : Data(data), Next(next)
+        {
+        }
+
+		Node(TNode&& data, Node* next) : Data(data), Next(next)
+		{
+		}
+
+        TNode Data;
+        Node* Next;
+    };
+    
+    struct Iterator
+    {
+        friend struct LinkedList<TNode>;
+        
+        Iterator(Node* pointer) : mPointer(pointer)
+        {
+        }
+        
+        bool operator!=(const Iterator& that) 
+        {
+            return that.mPointer != mPointer;
+        }
+        
+        const TNode& operator*() 
+        {
+            return mPointer->Data;
+        }
+        
+        void operator++() 
+        {
+            mPointer = mPointer->Next;
+        }
+        
+    private:
+        Node* mPointer;
+    };
+    
+public:
+        
+    LinkedList()
+    {
+        mHead = nullptr;
+    }
+    
+    LinkedList(const LinkedList& that)
+    {   
+        if(that.empty())
+        {
+            mHead = nullptr;  
+        }
+        else
+        {
+            mHead = new Node(that.mHead->Data, nullptr);            
+            auto current =  mHead;
+            
+            for(auto thatCurrent= that.mHead->Next; thatCurrent != nullptr; thatCurrent = thatCurrent->Next)
+            {
+                current->Next = new Node(thatCurrent->Data, nullptr);
+                current = current->Next;
+            }
+        }           
+    }
+    
+    LinkedList(LinkedList&& that)
+    {
+        mHead = that.mHead;
+        that.mHead = nullptr;        
+    }
+    
+    ~LinkedList()
+    {        
+        auto current = mHead;
+        while(current != nullptr)
+        {
+            auto tmp = current;
+            current = current->Next;
+            delete tmp;
+        }
+    }
+    
+    const Iterator begin() const
+    {
+        return Iterator(mHead);
+    }
+    
+    const Iterator end() const
+    {
+        return Iterator(nullptr);
+    }
+    
+    std::size_t size() const 
+    {
+        int size = 0;
+        
+        for(auto current = mHead; current != nullptr; current = current->Next)
+            ++size;
+        
+        return size;
+    }
+
+    bool empty() const
+    {
+        return mHead == nullptr;
+    }
+    
+    void push_front(const TNode& data)
+    {
+        mHead = new Node(data, mHead);
+    }
+
+	void push_front(TNode&& data)
+	{
+		// TODO: Writing this function feels wrong, is it?
```

Comment:
Sure. The indentation is wrong, for one :)
Secondly, it uses explicit 'new', which should be avoided.
And last, but not least: it won't work as expected. Try creating a list of non-copyable objects, for instance a list of std::unique_ptr. It won't work with this implementation, but it should.